Fix referencing of current requested procedure or pertinent other evidence

hackermd opened the pull request:
Closing #62.

Includes any data set provided via `evidence` into *Current Requested Procedure Evidence Sequence* if a given data set is referenced in the document content tree via a content item of value type IMAGE or COMPOSITE. If the data set is not referenced, it will be included in *Pertinent Other Evidence Sequence* if `record_evidence` is set to `True`.

CPBridge commented:
Haven't done a full review yet but one thing that I immediately noticed is that, unless I've missed something, it is still possible to construct an SR where a particular SOP instance is referenced in the content tree but not included in the evidence parameter. In this case highdicom will still not complain and produce an SR that is technically invalid. I think there needs to be an extra loop that checks whether all `ref_uids` are present in `evidence` and raises an exception of something is missing?

hackermd commented:
> Haven't done a full review yet but one thing that I immediately noticed is that, unless I've missed something, it is still possible to construct an SR where a particular SOP instance is referenced in the content tree but not included in the evidence parameter. In this case highdicom will still not complain and produce an SR that is technically invalid. I think there needs to be an extra loop that checks whether all ref_uids are present in evidence and raises an exception of something is missing?

Very good point! Will add the check

hackermd commented:
@CPBridge I added the check. The constructor will now raise a ValueError if any of the referenced data sets is not provided as `evidence`. Please note that I removed the requirement that all referenced data sets are part of the same study. As you pointed out in #62 that is probably too restrictive. I further noticed that the Current Requested Procedure Evidence Sequence, the Pertinent Other Evidence Sequence, and the Predecessor Documents Sequence can contain more than one item, which means that series and instances from multiple studies may be referenced.

CPBridge commented on src/highdicom/sr/sop.py:
```diff
@@ -180,67 +180,168 @@ def __init__(
+                raise ValueError(
+                    'Data sets provided as evidence must all belong to the '
+                    'same study.'
+                )
```
I think you maybe intended to remove this check and exception?

hackermd replied:
I did. Somehow it made it back in. Addressed with 8e6ad6259f7553cbd708bbc392d2b2f41df540ae

CPBridge commented on src/highdicom/sr/sop.py:
```diff
@@ -180,67 +180,168 @@ def __init__(
+        ref_uids = [
+            ref.ReferencedSOPSequence[0].ReferencedSOPInstanceUID
+            for ref in references
+        ]
```
Minor preference probably - but I would tend to make this a set instead of a list since order is unimportant, there may be duplicates references that only need to be in this container once, and it will be used for lots of look-up operations. That would also allow you to use set operations between `ref_uids` and `evd_uids` later, rather than another loop.

hackermd replied:
Addressed with 8e6ad6259f7553cbd708bbc392d2b2f41df540ae

CPBridge commented on src/highdicom/sr/sop.py:
```diff
@@ -180,67 +180,168 @@ def __init__(
+
+        Returns
+        -------
+        Tuple[List[pydicom.dataset.Dataset], List[pydicom.dataset.Dataset]]
```
As discussed previously, my preferred style is to list elements of the output tuple as individual named outputs as this leads to clearer docs in my opinion. Take it or leave it.

hackermd replied:
I agree and I think we should adopt this style of documenting the return value. However, I would like the same style to be consistently applied throughout the library and suggest updating docstrings via a separate PR.

CPBridge replied:
I did some grepping tosee whether there are in fact any other instances of this in the codebase, and I found a couple. Dumping here for reference:
- `highdicom.seg.content.DimensionIndexSequence.get_index_values`
- `highdicom.spatial.map_pixel_into_coordinate_system` and `highdicom.spatial.map_coordinate_into_pixel_matrix`. These may actually be better left as they are as I imagine the output will typically not be destructered

CPBridge commented on tests/test_sr.py:
```diff
@@ -2336,6 +2336,12 @@ def setUp(self):
+        assert len(ref_evd_items) == 1
+        with pytest.raises(AttributeError):
+            assert self.PertinentOtherEvidenceSequence
+
```
A test for evidence from different studies wouldn't be a bad idea?

hackermd replied:
Addressed with dbe4c7707a002b3c9e7698d2e5d509cff33a331f

CPBridge commented on src/highdicom/sr/sop.py:
```diff
@@ -180,67 +180,168 @@ def __init__(
+                ref_collection[key].append(evd_item)
+            else:
+                unref_collection[key].append(evd_item)
+            evd_uids.append(evd.SOPInstanceUID)
```
Is it worth checking that the item is not already in the sequence, in case the user passes duplicate items in `evidence`?

hackermd replied:
Addressed with 8e6ad6259f7553cbd708bbc392d2b2f41df540ae

CPBridge reviewed (changes requested):
A few further comments - nothing blocking merge except the vestigial study UID check. Also there seem to be some mypy failures

hackermd commented:
Addressed mypy failures with 58351d685307ffb0bebebdd2d1617c88c7f6823a

CPBridge reviewed (approved):
Looks good! Just one more flake8 bug to squash

hackermd commented:
> Looks good! Just one more flake8 bug to squash

Addressed with 7284f0c4f255ad654e232572e36a9013c078dbca
